File: 2015/12/12.py
from json import loads
import builtins


def parser(input): 
    return loads(input)


def solver(document):
    yield get_sum(document, True)
    yield get_sum(document, False)


def get_sum(document, allow_red):
    match type(document):
        case builtins.int:  return document
        case builtins.list: return sum(get_sum(subdocument , allow_red) for subdocument in document)
        case builtins.dict: 
            if allow_red or 'red' not in document.values():
                return get_sum(list(document.values()), allow_red) 
            else:
                return 0
        case _: return 0
        

File: 2015/12/12.rs
aoc::main!(); 

fn parser(input: &str) -> Value {
    input.parse().unwrap()
}

fn part_1(loads: Value) -> i64 {
    sum_json(&loads, &json!(""))
}

fn part_2(loads: Value) -> i64 {
    sum_json(&loads, &json!("red"))
}

fn sum_json(loads: &Value, ban_word: &Value) -> i64 {
    match loads {
        Value::Number(n) => n.as_i64().unwrap(),
        Value::Object(o) => if o.values().contains(&ban_word) {0} else {o.iter().map(|l| sum_json(l.1, ban_word)).sum()},
        Value::Array(a)  => a.iter().map(|l| sum_json(l, &ban_word)).sum(),
        Value::String(_) => 0,
        _ => unreachable!(),
    }
}